File: app/src/main/java/com/usharik/app/service/FirebaseAnalyticsService.java
package com.usharik.app.service;

import android.os.Bundle;

import com.google.firebase.analytics.FirebaseAnalytics;

import javax.inject.Inject;
import javax.inject.Singleton;

@Singleton
public class FirebaseAnalyticsService {

    public static final String KEY_BUTTON = "BUTTON";

    private final FirebaseAnalytics firebaseAnalytics;

    @Inject
    public FirebaseAnalyticsService(FirebaseAnalytics firebaseAnalytics) {
        this.firebaseAnalytics = firebaseAnalytics;
    }

    public void setCollectionEnabled(boolean enabled) {
        firebaseAnalytics.setAnalyticsCollectionEnabled(enabled);
    }

    public void logButtonClick(String eventName, String buttonName) {
        if (eventName == null || buttonName == null) {
            return;
        }
        Bundle bundle = new Bundle();
        bundle.putString(KEY_BUTTON, buttonName);
        firebaseAnalytics.logEvent(eventName, bundle);
    }

    public void logNextWord(String word) {
        Bundle bundle = new Bundle();
        bundle.putString("WORD", word);
        firebaseAnalytics.logEvent("NEXT_WORD", bundle);
    }

    public void logNextWordAction(String actionName) {
        Bundle bundle = new Bundle();
        bundle.putString("NEXT_WORD_ACTION", actionName);
        firebaseAnalytics.logEvent("NEXT_WORD_ACTION", bundle);
    }

    public void logMistake(Bundle bundle) {
        if (bundle == null || bundle.isEmpty()) {
            return;
        }
        firebaseAnalytics.logEvent("MISTAKE", bundle);
    }

    public void logSettings(boolean switchOffAnimation) {
        Bundle bundle = new Bundle();
        bundle.putBoolean("SWITCH_OFF_ANIMATION", switchOffAnimation);
        firebaseAnalytics.logEvent("SETTINGS", bundle);
    }

    public void logSingleCaseAnswer(boolean isCorrect, String selected, String correct, String word, String caseName) {
        Bundle bundle = new Bundle();
        bundle.putString("RESULT", isCorrect ? "CORRECT" : "INCORRECT");
        bundle.putString("SELECTED_ANSWER", selected);
        bundle.putString("CORRECT_ANSWER", correct);
        bundle.putString("WORD", word);
        bundle.putString("CASE", caseName);
        firebaseAnalytics.logEvent("SINGLE_CASE_ANSWER", bundle);
    }

    public void logSingleCaseNavigation(String buttonName, String word) {
        Bundle bundle = new Bundle();
        bundle.putString(KEY_BUTTON, buttonName);
        bundle.putString("WORD", word);
        firebaseAnalytics.logEvent("SINGLE_CASE_NAVIGATION", bundle);
    }

    public void logHandbookOpen() {
        Bundle bundle = new Bundle();
        bundle.putString("HANDBOOK_FRAGMENT", "OPEN");
        firebaseAnalytics.logEvent("HANDBOOK_FRAGMENT", bundle);
    }

    public void logEvent(String eventName, Bundle bundle) {
        if (eventName == null) {
            return;
        }
        firebaseAnalytics.logEvent(eventName, bundle);
    }
}

File: app/src/main/kotlin/com/usharik/app/service/FirebaseAnalyticsService.kt
package com.usharik.app.service

import android.os.Bundle
import com.google.firebase.analytics.FirebaseAnalytics

class FirebaseAnalyticsService(private val analytics: FirebaseAnalytics) {
    fun setCollectionEnabled(enabled: Boolean) = analytics.setAnalyticsCollectionEnabled(enabled)
    fun logButtonClick(event: String, button: String) = logEvent(event, Bundle().apply { putString(KEY_BUTTON, button) })
    fun logNextWord(word: String) = logEvent("NEXT_WORD", Bundle().apply { putString("WORD", word) })
    fun logNextWordAction(action: String) = logEvent("NEXT_WORD_ACTION", Bundle().apply { putString("NEXT_WORD_ACTION", action) })
    fun logMistake(bundle: Bundle?) { if (bundle?.isEmpty == false) analytics.logEvent("MISTAKE", bundle) }
    fun logSettings(switchOffAnimation: Boolean) = logEvent("SETTINGS", Bundle().apply { putBoolean("SWITCH_OFF_ANIMATION", switchOffAnimation) })
    fun logSingleCaseAnswer(correct: Boolean, selected: String, answer: String, word: String, caseName: String) = logEvent("SINGLE_CASE_ANSWER", Bundle().apply { putString("RESULT", if (correct) "CORRECT" else "INCORRECT"); putString("SELECTED_ANSWER", selected); putString("CORRECT_ANSWER", answer); putString("WORD", word); putString("CASE", caseName) })
    fun logSingleCaseNavigation(button: String, word: String) = logEvent("SINGLE_CASE_NAVIGATION", Bundle().apply { putString(KEY_BUTTON, button); putString("WORD", word) })
    fun logHandbookOpen() = logEvent("HANDBOOK_FRAGMENT", Bundle().apply { putString("HANDBOOK_FRAGMENT", "OPEN") })
    fun logDailyReminderShown(streak: Int, words: Int, exercises: Int) = logEvent("daily_reminder_shown", Bundle().apply { putInt("inactivity_streak", streak); putInt("words_completed_yesterday", words); putInt("exercises_completed_yesterday", exercises) })
    fun logEvent(event: String, bundle: Bundle? = null) { analytics.logEvent(event, bundle) }
    companion object { const val KEY_BUTTON = "BUTTON" }
}
